### PR TITLE
[Mono]: Reduce Mono AOT cross compiler x64 memory footprint.

### PR DESCRIPTION
**Pull request by lateralusX**
Building .net8 S.P.C using Mono AOT cross compiler in full AOT consumes a large amount of memory (up to 6 GB). This is mainly due to generated LLVM module not being optimized at all while kept in memory during full module generation. Mono x64 also lacks support for several intrinsics as well as Vector 256/512 that in turn leads to massive inlining of intrinsics functions generating a very large LLVM module, where majority of this code ends up as dead code due to IsSupported/IsHardwareAccelerated returning false.

The follow commit adjusts several things that will bring down the memory usage, compiling .net8/.net9 Mono S.P.C on x64 Windows from 6 GB down to ~750 MB.

* Use PSNE implementations on intrinsics not supported on Mono.
* Add ILLinker substitutions for intrinsics not supported on Mono. Enables ILLinker to do dead code elimination, reduce code to AOT compile.
* Prevent aggressive inlining for a couple of unsupported intrinsics types making sure we don't end up with excessive inlining, exploding code size.
* Run a couple of LLVM optimization passes on each generated method doing early code simplification and dead code elimination during LLVM module generation.
* Explicit SN_get_IsHardwareAccelerated/SN_get_IsSupported intrinsics implementation for all unsupported Mono x64 SIMD intrinsics.
* Fixed numerous memory leaks in Mono AOT cross compiler code.
* Fix a couple of sequence points free after use errors.
* Fix an anonymous struct build warning triggering build error for LLVM enabled cross compiler on Windows.

### CODE REVIEW

File: src/mono/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Intrinsics.Vectors.xml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<linker>
+  <assembly fullname="System.Private.CoreLib">
+    <type fullname="System.Runtime.Intrinsics.Vector256">
```
Comment (marek-safar) [inline src/mono/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Intrinsics.Vectors.xml:3]: Is this true for all codegens (e.g. interpreter)

File: src/mono/mono/mini/aot-compiler.c [hunk 2]
```diff
@@ -4208,22 +4208,24 @@ get_plt_entry (MonoAotCompile *acfg, MonoJumpInfo *patch_info)
...
-			res->symbol = g_strdup_printf ("%s_%d", res->symbol, synchronized_symbol_idx);
-			if (res->debug_sym)
-				res->debug_sym = g_strdup_printf ("%s_%d", res->debug_sym, synchronized_symbol_idx);
+			char *tmp_symbol = res->symbol;
+			res->symbol = mono_mempool_strdup_printf (acfg->mempool, "%s_%d", res->symbol, synchronized_symbol_idx);
+			g_free (tmp_symbol);
+			if (res->debug_sym) {
+				char *tmp_debug_sym = res->debug_sym;
+				res->debug_sym = mono_mempool_strdup_printf (acfg->mempool, "%s_%d", res->debug_sym, synchronized_symbol_idx);
+				g_free (tmp_debug_sym);
```
Comment (lambdageek) [inline src/mono/mono/mini/aot-compiler.c:4217]: `get_plt_entry_debug_sym` calls `sanitize_symbol` which calls `mono_mempool_strdup`, so it's not ok to always call  `g_free (tmp_debug_sym)`

Comment (lateralusX) [reply src/mono/mono/mini/aot-compiler.c:4217]: Interesting, this is upstreamed from .net8 implementation so maybe that implementation changed, need to investigate and adjust that path.

**Comment by vargaz** [comment]
Wouldn't be better to split this into smaller PRs ?

**Comment by lambdageek** [comment]
I wonder if we can make working with symbols a little more typesafe so that we have some distinction between a mempool allocated symbol and a temporary malloc-allocated symbol.  Maybe we can just pass around a `GString` for the temporary ones?

File: src/mono/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Intrinsics.Vectors.xml (hunk 1 above)
Comment (lateralusX) [reply src/mono/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Intrinsics.Vectors.xml:3]: Looks like interpreter mark all vectors as not being hardware accelerated:

```
} else if (in_corlib &&
		   (!strncmp ("System.Runtime.Intrinsics", klass_name_space, 25) &&
			!strncmp ("Vector", klass_name, 6) &&
			!strcmp (tm, "get_IsHardwareAccelerated"))) {
	*op = MINT_LDC_I4_0;
}
```

so for that case it should be ok.

File: src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems [hunk 3]
```diff
@@ -2574,18 +2574,27 @@
...
-    <Compile Include="$(MSBuildThisFileDirectory)System\Runtime\Intrinsics\X86\Aes.cs" />
-    <Compile Include="$(MSBuildThisFileDirectory)System\Runtime\Intrinsics\X86\Avx.cs" />
-    <Compile Include="$(MSBuildThisFileDirectory)System\Runtime\Intrinsics\X86\Avx2.cs" />
-    <Compile Include="$(MSBuildThisFileDirectory)System\Runtime\Intrinsics\X86\Avx512BW.cs" />
-    <Compile Include="$(MSBuildThisFileDirectory)System\Runtime\Intrinsics\X86\Avx512CD.cs" />
-    <Compile Include="$(MSBuildThisFileDirectory)System\Runtime\Intrinsics\X86\Avx512DQ.cs" />
-    <Compile Include="$(MSBuildThisFileDirectory)System\Runtime\Intrinsics\X86\Avx512F.cs" />
-    <Compile Include="$(MSBuildThisFileDirectory)System\Runtime\Intrinsics\X86\Avx512Vbmi.cs" />
-    <Compile Include="$(MSBuildThisFileDirectory)System\Runtime\Intrinsics\X86\AvxVnni.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)System\Runtime\Intrinsics\X86\Aes.cs"/>
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems:2577]: nit: the diff here is unnecessarily large as some lines had the whitespace removed at the end of the line between the quote and the close of the tag.

Itd be easier to review if unmodified lines were preserved (and it'd be ideal to keep the space even in modified lines for consistency with the rest of the file)

File: src/mono/mono/mini/aot-compiler.c (hunk 2 above)
Comment (lateralusX) [reply src/mono/mono/mini/aot-compiler.c:4217]: OK, will drop that g_free, since changes done in get_plt_entry_debug_sym makes sure to free any heap allocated memory  if we end up returning a mempool allocated string.

File: src/mono/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Intrinsics.x86.xml [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    <type fullname="System.Runtime.Intrinsics.X86.Avx512Vbmi/X64">
+      <method signature="System.Boolean get_IsSupported()" body="stub" value="false" />
+    </type>
+    <type fullname="System.Runtime.Intrinsics.X86.AvxVnni">
+      <method signature="System.Boolean get_IsSupported()" body="stub" value="false" />
+    </type>
+    <type fullname="System.Runtime.Intrinsics.X86.AvxVnni/X64">
+      <method signature="System.Boolean get_IsSupported()" body="stub" value="false" />
+    </type>
+    <type fullname="System.Runtime.Intrinsics.X86.Fma">
```
Comment (tannergooding) [inline src/mono/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Intrinsics.x86.xml:66]: Is this missing `Bmi1`/`Bmi2` which also use the VEX encoding?

File: src/mono/mono/mini/simd-intrinsics.c [hunk 5]
```diff
@@ -1416,9 +1430,16 @@ emit_sri_vector (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSignature *fsi
 	else
 		return NULL;
 
-	if (vector_size == 256 || vector_size == 512)
-		return NULL; 
-		
+	if (vector_size == 256 || vector_size == 512) {
+		if (id == SN_get_IsHardwareAccelerated ) {
```
Comment (tannergooding) [inline src/mono/mono/mini/simd-intrinsics.c:1434]: For RyuJIT, we have a fallback that handles any unrecognized `get_IsHardwareAccelerated` and `get_IsSupported` APIs for `System.Numerics`, `System.Runtime.Intrinsics`, and `System.Runtime.Intrinsics.*` to ensure they can be treated as `constant false`. Each namespace is being handled a little bit differently since `get_IsSupported` for some namespaces needs to fallback to user-code instead.

Would it be a good idea to similarly make this general-purpose. Notably `vector_size == 64` should also be `false` on `x86/x64` for example?

File: src/mono/mono/mini/simd-intrinsics.c [hunk 6]
```diff
@@ -1416,9 +1430,16 @@ emit_sri_vector (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSignature *fsi
 	else
 		return NULL;
 
-	if (vector_size == 256 || vector_size == 512)
-		return NULL; 
-		
+	if (vector_size == 256 || vector_size == 512) {
```
Comment (tannergooding) [inline src/mono/mono/mini/simd-intrinsics.c:1433]: Should `IsSupported` be handled for `Vector64/256/512` since the set of types is well-known and the same as for `Vector128`, that would allow it to be trivially constant true/false as well.

File: src/mono/mono/mini/simd-intrinsics.c [hunk 7]
```diff
@@ -5911,8 +5981,14 @@ static
...
 {
+	if (!strcmp (class_ns, "System.Runtime.Intrinsics.X86")) {
+		return emit_hardware_intrinsics (cfg, cmethod, fsig, args,
+			supported_x86_intrinsics, sizeof (supported_x86_intrinsics),
+			emit_x86_intrinsics);
+	}
+
 	if (!strcmp (class_ns, "System.Runtime.Intrinsics")) {
-		if (!strcmp (class_name, "Vector128"))
+		if (!strcmp (class_name, "Vector128") || !strcmp(class_name, "Vector256") || !strcmp(class_name, "Vector512"))
```
Comment (tannergooding) [inline src/mono/mono/mini/simd-intrinsics.c:5991]: What about `Vector64`?

File: src/mono/mono/mini/simd-intrinsics.c [hunk 8]
```diff
@@ -5911,8 +5981,14 @@ static
 MonoInst*
 arch_emit_simd_intrinsics (const char *class_ns, const char *class_name, MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSignature *fsig, MonoInst **args)
 {
+	if (!strcmp (class_ns, "System.Runtime.Intrinsics.X86")) {
```
Comment (tannergooding) [inline src/mono/mono/mini/simd-intrinsics.c:5984]: Is there a similar line for `System.Runtime.Intrinsics.Arm` and `System.Runtime.Intrinsics.Wasm` missing?

File: src/mono/mono/mini/simd-intrinsics.c [hunk 9]
```diff
@@ -6118,11 +6211,36 @@ mono_simd_decompose_intrinsic (MonoCompile *cfg, MonoBasicBlock *bb, MonoInst *i
...
+	/*
+	* If a method has been marked with aggressive inlining, check if we support
+	* aggressive inlining of the intrinsics type, if not, ignore aggressive inlining
+	* since it could end up inlining a large amount of code that most likely will end
+	* up as dead code.
+	*/
+	if (!strcmp (m_class_get_name_space (cmethod->klass), "System.Runtime.Intrinsics")) {
+		if (!strncmp(m_class_get_name (cmethod->klass), "Vector", 6)) {
+			const char *vector_type = m_class_get_name (cmethod->klass) + 6;
+			if (!!strcmp(vector_type, "256`1") || !strcmp(vector_type, "512`1"))
```
Comment (tannergooding) [inline src/mono/mono/mini/simd-intrinsics.c:6227]: `Vector64`?

File: src/mono/mono/mini/simd-intrinsics.c (hunk 8 above)
Comment (lateralusX) [reply src/mono/mono/mini/simd-intrinsics.c:5984]: Handled elsewhere in that source file.

Comment (lateralusX) [reply src/mono/mono/mini/simd-intrinsics.c:5984]: You will find Arm and Wasm under respective defines.

File: src/mono/mono/mini/simd-intrinsics.c [hunk 10]
```diff
@@ -6118,11 +6211,36 @@ mono_simd_decompose_intrinsic (MonoCompile *cfg, MonoBasicBlock *bb, MonoInst *i
...
+
+gboolean
+mono_simd_unsupported_aggressive_inline_intrinsic_type (MonoMethod *cmethod)
+{
+	/*
+	* If a method has been marked with aggressive inlining, check if we support
+	* aggressive inlining of the intrinsics type, if not, ignore aggressive inlining
+	* since it could end up inlining a large amount of code that most likely will end
+	* up as dead code.
+	*/
```
Comment (tannergooding) [inline src/mono/mono/mini/simd-intrinsics.c:6223]: Could you help me understand why this is needed on Mono a bit?

The software fallback for V128/V256/V512 is currently implemented as doing 2x operations on the lower/upper halves (except for a couple of methods like `Shuffle` which operate on the full vector). So `V512` is implemented as `2x V256`, `V256` is implemented as `2x V128`, and `V128` is implemented as `2x V64`. `V64` is then implemented as a loop over the scalar elements with potentially large amounts of generic code.

So I would imagine that on any platform where `V128` is supported, that the codegen for `V256`/`V512` should be generally nice/small, even if aggressively inlined and with no real dead code elimination required. Even in the case where you have an unsupported type like `V256<Guid>`, the first `V128` operation should have a `PNSE` thrown (since whether a type is supported or not is typically a known constant).

So I'd only expect that this is needed for V64 on x86/x64 where there is no acceleration and it has to hit the scalar fallback with the loop over the elements. For RyuJIT this loop isn't an issue due to the generic specialization we do, allowing us to get it down to only the code path that is actually used. I believe this isn't possible for Mono today and is non-trivial to add, so the skipping of inlining does make sense in that regard.

File: src/mono/mono/mini/simd-intrinsics.c (hunk 8 above)
Comment (tannergooding) [reply src/mono/mono/mini/simd-intrinsics.c:5984]: Do they not need a path to ensure that `IsSupported` returns constant `false` and the intrinsics directly generate a `PNSE` exception, rather than hitting the recursive fallback, or is that handled elsewhere as well?

File: src/mono/mono/mini/simd-intrinsics.c (hunk 10 above)
Comment (lateralusX) [reply src/mono/mono/mini/simd-intrinsics.c:6223]: So without disabling aggressive inlining I ended up with methods that where huge, and not doing aggressive inline on these types (but still honor inline limit size) made them sane again. I will need to re-iterate around that change in order to tell exactly what happened with our inliner in the aggressive case.

File: src/mono/mono/mini/simd-intrinsics.c (hunk 8 above)
Comment (lateralusX) [reply src/mono/mono/mini/simd-intrinsics.c:5984]: I mainly focused on x86 in this PR, other code will still go through the fallbacks, but could probably be enhanced as well.

File: src/mono/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Intrinsics.Vectors.xml (hunk 1 above)
Comment (lewing) [reply src/mono/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Intrinsics.Vectors.xml:3]: cc @kg

Comment (kg) [reply src/mono/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Intrinsics.Vectors.xml:3]: We mark v128 as hardware accelerated in interp in some cases, I believe.

**Comment by steveisok** [comment]
`System.Collections.Concurrent` failed to AOT in a couple of suites. Not sure why, but here's the log https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-97096-merge-21c2561d7c2b483faf/Invariant.Tests/1/console.3bf63a3a.log?helixlogtype=result

File: src/mono/mono/mini/simd-intrinsics.c (hunk 9 above)
Comment (fanyang-mono) [reply src/mono/mono/mini/simd-intrinsics.c:6227]: Extra exclamation mark in front of the check for Vector256.

File: src/mono/mono/mini/aot-runtime.c [hunk 11]
```diff
@@ -3390,10 +3390,13 @@ decode_exception_debug_info (MonoAotModule *amodule,
 			MonoJitMemoryManager *jit_mm = get_default_jit_mm ();
 			jit_mm_lock (jit_mm);
 			/* This could be set already since this function can be called more than once for the same method */
-			if (!g_hash_table_lookup (jit_mm->seq_points, method))
```
Comment (vargaz) [inline src/mono/mono/mini/aot-runtime.c:3393]: This appears unrelated.

File: src/mono/mono/mini/exceptions-amd64.c [hunk 12]
```diff
@@ -1044,14 +1044,14 @@ mono_arch_unwindinfo_add_push_nonvol (PUNWIND_INFO unwindinfo, MonoUnwindOp *unw
 
 	codeindex = MONO_MAX_UNWIND_CODES - (++unwindinfo->CountOfCodes);
 	unwindcode = &unwindinfo->UnwindCode [codeindex];
-	unwindcode->UnwindOp = UWOP_PUSH_NONVOL;
```
Comment (vargaz) [inline src/mono/mono/mini/exceptions-amd64.c:1047]: These are unrelated.

File: src/mono/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Intrinsics.Vectors.xml (hunk 1 above)
Comment (lateralusX) [reply src/mono/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Intrinsics.Vectors.xml:3]: ok, so then this should be OK since it only affects v256/v512.

File: src/mono/mono/mini/exceptions-amd64.c (hunk 12 above)
Comment (lateralusX) [reply src/mono/mono/mini/exceptions-amd64.c:1047]: Yes, this is fixing a build error on Windows that I fixed as part of this PR (since I implemented the change on Windows, so it was in the way). If it make sense it could be moved into a separate PR or kept in this one, for me it doesn't matter.

File: src/mono/mono/mini/aot-compiler.c (hunk 2 above)
Comment (lateralusX) [reply src/mono/mono/mini/aot-compiler.c:4217]: I changed `sanitize_symbol `to use `g_strdup` instead of `mempool` since there is no clear ownership of the passed in string and what "type" of allocation that gets returned back to caller. Since the majority of the current implementation already used none mempool allocated strings it made sense to just move over to use `g_strdup` instead of `mempool` and keep the free as implemented in above code.

File: src/mono/mono/mini/simd-intrinsics.c (hunk 10 above)
Comment (lateralusX) [reply src/mono/mono/mini/simd-intrinsics.c:6223]: Just validated this, with disabling the aggressive inlining as above a .net9 full AOT of S.P.C takes ~1.7 GB of memory, and not doing this will consume an additional 600 MB of memory, so I believe its worth preventing aggressive inlining for these types that are not hardware accelerated on any of the Mono supported platforms. I didn't add V64 since it seems to be handled a little differently on at least ARM case. I won't have bandwidth at the moment to do more deep analysis around why aggressive inlining of these template types cause that large increase in memory so I think doing this change, at least short term is worth it, we could probably file an issue around the bloat of Vector256`1 and Vector512`1, but maybe the better long term solution is to actually implement intrinsic support for these types.

File: src/mono/mono/mini/simd-intrinsics.c (hunk 5 above)
Comment (lateralusX) [reply src/mono/mono/mini/simd-intrinsics.c:1434]: Reverted this change to rely on fallback for get_IsHardwareAccelerated.

File: src/mono/mono/mini/simd-intrinsics.c (hunk 10 above)
Comment (BrzVlad) [reply src/mono/mono/mini/simd-intrinsics.c:6229]: For interpreter I'm handling this in a general fashion by detecting early dead pieces of code and not inlining any of the calls there. https://github.com/dotnet/runtime/pull/97514. It is possible that a similar approach for jit can produce further improvements without having to special case classes.

Comment (vargaz) [reply src/mono/mono/mini/simd-intrinsics.c:6229]: Trying to reenable branch optimizations when using llvm in this pr:
https://github.com/dotnet/runtime/pull/97189

**Comment by vitek-karas** [comment]
@MichalStrehovsky FYI for the substitution changes

**Comment by lewing** [comment]
Hopfully I resolved the conflicts correctly, someone should review.

**Review by vargaz — APPROVED** [review]

**Comment by vargaz** [comment]
Failures are unrelated.

**Comment by matouskozak** [comment]
This PR looks to be responsible for ~200kB package size improvement on iOS HelloWorld (range of commits https://github.com/dotnet/runtime/compare/339443b0408f45503380558aa2cf2a9ecc2cbaa9...a79c62ddc8089cf2879ed36eac9aa333b32bde5f) 
image.

Good job everyone!

**Comment by matouskozak** [comment]
> This PR looks to be responsible for ~200kB package size improvement on iOS HelloWorld (range of commits [339443b...a79c62d](https://github.com/dotnet/runtime/compare/339443b0408f45503380558aa2cf2a9ecc2cbaa9...a79c62ddc8089cf2879ed36eac9aa333b32bde5f)) image.
> 
> Good job everyone!

Edit. after the subsequent fix to this PR (https://github.com/dotnet/runtime/pull/98515), the package size improvements on iOS HelloWorld are mostly gone (i.e., back to original values).
